### ISSUE TITLE
dynamic cooldowns

### ISSUE THREAD

**Issue by EEKIM10** [Rapptz/discord.py#2127]
### The Problem

unless im an idot, there is no current ways to have "dynamic" cooldowns. by this i mean for example, if x has y role, give a cooldown of 10s, otherwise set a 20s cooldown.

### The Ideal Solution

perhaps a "check" kwarg, where you can do something like below:
```py
def cc(ctx):
    if ctx.author.id in [id_1, id_2]:
        return 10  # seconds to have on cooldown
    else:
        return 20

@commands.command()
@commands.cooldown(1, self.cc(ctx), ...)
```

### The Current Solution

the only current way i could imagine this is returning the cooldown *in* a custom check or something, not really sure

### Summary

basically a dynamically adjusted cooldown based on a custom check, like wait_for

**Comment by Vexs** [comment]
You can kind of implement this by leveraging command errors.
```py
    @command_name.error
    async def command_name_error(self, ctx, error):
        if isinstance(error,commands.CommandOnCooldown):
            if some_role in ctx.author.roles and error.retry_after < 10:
                if await ctx.bot.can_run(ctx):
                    await ctx.reinvoke()
                return
            await ctx.send(error)
```
I can't remember if can_run checks over cooldowns, if it does this obviously won't work and would be more annoying to implement.

**Comment by EEKIM10** [comment]
Actually, I haven’t tested it, but you could just make a custom function that returns a cooldown
```py
def cc(ctx):
    if local_role in ctx.author.roles:
        return commands.cooldown(...)
    else: 
        return commands.cooldown(...)
```
However it would have its issues, like you can’t pass ctx in a decorator

**Comment by Vexs** [comment]
That's... not going to work. The decorators aren't re-added each invoke.
